test/e2e/reconciler/ingress/controller_test.go

```diff
@@ -0,0 +1,612 @@
+
+	if deadline, ok := c.t.Deadline(); ok {
+		deadlinedCtx, deadlinedCancel := context.WithDeadline(ctx, deadline.Add(-10*time.Second))
+		ctx = deadlinedCtx
+		c.t.Cleanup(deadlinedCancel) // this does not really matter but govet is upset
+	}
+	ctx = ctx
+	cleanupCtx, cleanupCancel := context.WithCancel(context.Background())
+	c.t.Cleanup(func() {
+		c.t.Log("cleanup: ending ingress controller")
+		cancel()
+		<-cleanupCtx.Done()
+	})
+
+	cmd := exec.CommandContext(ctx, "ingress-controller", []string{
+		"-kubeconfig=" + c.kubeconfigPath,
+		"-envoyxds",
+		"-envoy-listener-port=" + c.envoyListenPort,
+		"-envoyxds-port=" + c.xdsListenPort,
+	}...)
+
+	c.t.Logf("running: %v", strings.Join(cmd.Args, " "))
+	logFile, err := os.Create(filepath.Join(c.artifactDir, "ingress-controller.log"))
+	if err != nil {
+		cleanupCancel()
+		return fmt.Errorf("could not create log file: %w", err)
+	}
+	log := bytes.Buffer{}
+	writers := []io.Writer{&log, logFile}
+	mw := io.MultiWriter(writers...)
+	cmd.Stdout = mw
+	cmd.Stderr = mw
+	if err := cmd.Start(); err != nil {
+		cleanupCancel()
+		return err
+	}
+	go func() {
+		defer func() { cleanupCancel() }()
+		err := cmd.Wait()
+		if err != nil && ctx.Err() == nil {
+			c.t.Errorf("`ingress-controller` failed: %w", err)
+		}
+	}()
+	return nil
+}
+
+func installService(ctx context.Context, server framework.RunningServer) error {
+	cfg, err := server.Config()
+	if err != nil {
+		return err
+	}
+	sourceClusterName, err := detectClusterName(cfg, ctx, "ingresses.networking.k8s.io")
+	if err != nil {
+		return fmt.Errorf("failed to detect cluster name: %w", err)
+	}
+	clients, err := kubernetesclientset.NewClusterForConfig(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to construct client for server: %w", err)
+	}
+	client := clients.Cluster(sourceClusterName)
+	_, err = client.CoreV1().Services(testNamespace).Create(ctx, &corev1.Service{
+		TypeMeta: metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: existingServiceName,
+			Labels: map[string]string{
+				"kcp.dev/cluster": clusterName,
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Name:     "http",
+					Port:     80,
+					Protocol: corev1.ProtocolTCP,
+				},
+			},
+			Selector: map[string]string{
+				"app": existingServiceName,
+			},
+		},
+		Status: corev1.ServiceStatus{},
+	}, metav1.CreateOptions{})
+	return err
+}
+
+func installNamespace(ctx context.Context, server framework.RunningServer) error {
+	cfg, err := server.Config()
+	if err != nil {
+		return err
+	}
+	clusterName, err := detectClusterName(cfg, ctx, "ingresses.networking.k8s.io")
+	if err != nil {
+		return fmt.Errorf("failed to detect cluster name: %w", err)
+	}
+	clients, err := kubernetesclientset.NewClusterForConfig(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to construct client for server: %w", err)
+	}
+	client := clients.Cluster(clusterName)
+	_, err = client.CoreV1().Namespaces().Create(ctx, &corev1.Namespace{
```

Comment:
Technically this can be broken-ish unless we wait for the RBAC caches to warm up, which I think we can do either with a SAR or with a `get` - see: 

https://github.com/openshift/ci-tools/blob/b85064b7d8d2dfc825555aeb815ddd8a74958ee0/cmd/ci-operator/main.go#L1089-L1108

@sttts is there any helper / best practice for this?

Reply:
am not aware of a helper, but could be there is one. Just wait.Until checking for 403.